include/qore/ExceptionSink.h

```diff
@@ -309,30 +331,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
     xsink->raiseException("OBJECT-ALREADY-DELETED", "attempt to access an already-deleted object of class '%s'", cname);
 }
 
-// returns a custom Qore program location for external modules to generate runtime exceptions with the source location
+//! returns a custom Qore program location for external modules to generate runtime exceptions with the source location
 class QoreExternalProgramLocationWrapper {
 public:
+    //! empty constructor; use set() to set the location
     DLLEXPORT QoreExternalProgramLocationWrapper();
 
+    //! copy ctor
+    DLLEXPORT QoreExternalProgramLocationWrapper(const QoreExternalProgramLocationWrapper&);
+
+    //! move ctor
+    DLLEXPORT QoreExternalProgramLocationWrapper(QoreExternalProgramLocationWrapper&&);
+
+    //! constructor setting the source location
     DLLEXPORT QoreExternalProgramLocationWrapper(const char* file, int start_line, int end_line,
-        const char* source = nullptr, int offset = 0);
+        const char* source = nullptr, int offset = 0, const char* lang = nullptr);
 
+    //! destructor; frees memory
     DLLEXPORT ~QoreExternalProgramLocationWrapper();
 
+    //! sets the program source location
     DLLEXPORT void set(const char* file, int start_line, int end_line,
-        const char* source = nullptr, int offset = 0);
+        const char* source = nullptr, int offset = 0, const char* lang = nullptr);
 
+    //! returns the source location
     DLLEXPORT const QoreProgramLocation& get() const {
         return *loc;
     }
 
+    //! returns the file name
+    DLLEXPORT const char* getFile() const {
+        return file_str.c_str();
+    }
+
+    //! returns the source
+    DLLEXPORT const char* getSource() const {
+        return source_str.c_str();
+    }
+
+    //! returns the language
+    DLLEXPORT const char* getLanguage() const {
+        return lang_str.c_str();
+    }
+
+    //! returns the start line
+    DLLEXPORT int getStartLine() const;
+
+    //! returns the start line
+    DLLEXPORT int getEndLine() const;
+
 private:
     // save strings for exceptions in case they are epheremal when this object is created
     std::string file_str;
     std::string source_str;
+    std::string lang_str;
 
     // actual exception location
     QoreProgramLocation* loc;
 };
 
+//! Stack location element abstract class
+/** @since %Qore 0.9
+*/
+class QoreStackLocation {
```

Comment:
this class looks like an implementation of a linked list, maybe we can use std::list for it and have QoreStackLocation as a template class for it holding only info about the current Program, statement etc, not holding the list directly.

Reply:
The reason it's implemented like this with a virtual function to return the next stack location is so that language modules (ex `jni`) can put a single placeholder on the stack, and then if and when an exception is raised, the object can populate all the stack entries in the language stack, the number of which entries will only be known at runtime in case of an exception.

The function returning the next ptr looks like this:
```
const QoreStackLocation* QoreJniStackLocationHelper::getNext() const {
    checkInit();
    assert(current < size);
    return (++current < size) ? this : stack_next;
}
```

So basically the "substack" of Java stack entries is created on demand if the stack is actually traversed due to an exception.